src/rust/engine/fs/src/snapshot.rs

```diff
@@ -119,22 +116,69 @@ least one path component, but at least one did not"
         directory.set_directories(protobuf::RepeatedField::from_vec(dirs));
         directory.set_files(protobuf::RepeatedField::from_vec(files));
         let digest = store.record_directory(&directory)?;
-        Ok(Snapshot {
-          fingerprint: digest.0,
-          digest: Some(digest),
-          path_stats: path_stats,
-        })
+        Ok(Snapshot { digest, path_stats })
       })
       .to_boxed()
   }
+
+  // TODO: How much should we be aiming to parallelise this?
+  // The old method returned a Future which did all of the work in one work-unit, on the PosixFS
+  // CpuPool. Is that sufficient? Or should we be kicking off each LMDB read in parallel on the
```

Comment:
> Or should we be kicking off each LMDB read in parallel on the pool?

This. Although I suspect it is already faster, if it's not (due to overhead in `CpuPool`), it [will be soon](https://github.com/alexcrichton/futures-rs/issues/411#issuecomment-341751574). As mentioned elsewhere: Store should hold a pool reference and expose futures.